File: backend/shared/models/billing.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

const (
	SubscriptionStatusActive    = "active"
	SubscriptionStatusCancelled = "cancelled"
	SubscriptionStatusPaused    = "paused"
)

type Plan struct {
	ID                  primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	Name                string             `bson:"name" json:"name"`
	BillingCycleOptions PlanCycleOptions   `bson:"billingCycleOptions" json:"billingCycleOptions"`
	Limits              PlanLimits         `bson:"limits" json:"limits"`
	Default             bool               `bson:"default" json:"default"`
}

type PlanLimits struct {
	MaxEvents              int `bson:"maxEvents" json:"maxEvents"`
	MaxMonthlyResponses    int `bson:"maxMonthlyResponses" json:"maxMonthlyResponses"`
	MaxMonthlyPipelineRuns int `bson:"maxMonthlyPipelineRuns" json:"maxMonthlyPipelineRuns"`
}

type PlanCycleOptions struct {
	Cycle string  `bson:"cycle" json:"cycle" oneOf:"monthly six-months yearly"`
	Price float32 `bson:"price" json:"price"`
}

type Subscription struct {
	ID                       primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	UserID                   primitive.ObjectID `bson:"userId" json:"userId"`
	PlanID                   primitive.ObjectID `bson:"planId" json:"planId"`
	StartDate                time.Time          `bson:"startDate" json:"startDate"`
	EndDate                  time.Time          `bson:"endDate" json:"endDate"`
	Status                   string             `bson:"status" json:"status" oneOf:"active cancelled paused"`
	BillingCycle             string             `bson:"billingCycle" json:"billingCycle" oneOf:"monthly six-months yearly"`
	NextUtilizationResetDate time.Time          `bson:"nextUtilizationResetDate" json:"nextUtilizationResetDate"`
	Utilization              Utilization        `bson:"utilization" json:"utilization"` // will need cron job to reset this scheduled to run every 24h
	Limits                   PlanLimits         `bson:"limits" json:"limits"`
}

type Utilization struct {
	EventsCreated int `bson:"eventsCreated" json:"eventsCreated"`
	Responses     int `bson:"responses" json:"responses"`
	PipelineRuns  int `bson:"pipelineRuns" json:"pipelineRuns"`
}

File: website/src/types/models/Billing.ts
export type SubscriptionStatus = 'active' | 'cancelled' | 'paused';

export const SUBSCRIPTION_STATUS_VALUES: Record<SubscriptionStatus, string> = {
  active: 'Active',
  cancelled: 'Cancelled',
  paused: 'Paused',
};

export type PlanCycle = 'monthly' | 'six-months' | 'yearly';

export const PLAN_CYCLE_VALUES: Record<PlanCycle, string> = {
  monthly: 'Monthly',
  'six-months': 'Six Months',
  yearly: 'Yearly',
};

export type ObjectId = string;

export type Plan = {
  id?: ObjectId;
  name: string;
  billingCycleOptions: PlanCycleOptions;
  limits: PlanLimits;
  default: boolean;
};

export type PlanLimits = {
  maxEvents: number;
  maxMonthlyResponses: number;
  maxMonthlyPipelineRuns: number;
};

export type PlanCycleOptions = {
  cycle: PlanCycle;
  price: number;
};

export type Subscription = {
  id?: ObjectId;
  userId: ObjectId;
  planId: ObjectId;
  startDate: Date;
  endDate: Date;
  status: SubscriptionStatus;
  billingCycle: PlanCycle;
  nextUtilizationResetDate: Date;
  utilization: Utilization;
  limits: PlanLimits;
};

export type Utilization = {
  eventsCreated: number;
  responses: number;
  pipelineRuns: number;
};
